File: stringio/pas/Str2num.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

const
  digits: array[0..15] of Char = '0123456789ABCDEF';

function Digit2index(digit: Char): Byte;

var
  index: Byte;

begin
  digit := UpCase(digit);
  index := 15;
  While (index > 0) and (digit <> digits[index]) do Dec(index);
  Digit2index := Index;
end;

function position_value(position,base: Byte): Longint;

var
  value: Longint;
  index: Byte;

begin
  value := 1;
  For index := 2 to position do value := value*base;
  position_value := value;
end;

function Str2num(str: String; base: Byte): Longint;

var
  value: Longint;
  index: Byte;

begin
  value := 0;
  If (base in [2,10,16]) then
    For index := 1 to Length(str) do
      Inc(value,Digit2index(str[index])*
                position_value(Length(str)-index+1,base));
  Str2num := value;
end;

File: stringio/Str2num.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

int32_t Str2num (const String *str, uint8_t base) {
  int32_t value = 0;

  if ((base == 2) || (base == 10) || (base == 16)) {
    size_t len = Length (str);
    const unsigned char *s = GetStr (str);

    while (len) {
      unsigned char c = UpCase (*s);

      s++;

      if ((c >= '0') && (c <= '9'))
        c -= '0';
      else if ((c >= 'A') && (c <= 'F'))
        c -= 'A' - 10;
      else
        c = 0;
      if (c >= base)
        c = 0;

      value = value * base + c;
      len--;
    }
  }

  return value;
}
